tapdb/sqlc/migrations/000012_multiverse_tree.up.sql

```diff
@@ -0,0 +1,83 @@
+CREATE TABLE IF NOT EXISTS multiverse_roots (
+    id BIGINT PRIMARY KEY,
+
+    -- For the namespace root, we set the foreign key constraint evaluation to
+    -- be deferred until after the database transaction ends. Otherwise, if the
+    -- root of the SMT is deleted temporarily before inserting a new root, then
+    -- this constraint is violated as there's no longer a root that this
+    -- universe tree can point to.
+    namespace_root VARCHAR UNIQUE NOT NULL REFERENCES mssmt_roots(namespace) DEFERRABLE INITIALLY DEFERRED,
+
+    -- This field is an enum representing the proof type stored in the given
+    -- universe.
+    proof_type TEXT NOT NULL CHECK(proof_type IN ('issuance', 'transfer'))
+);
+
+CREATE TABLE IF NOT EXISTS multiverse_leaves (
+    id BIGINT PRIMARY KEY,
+
+    multiverse_root_id BIGINT NOT NULL REFERENCES multiverse_roots(id),
+
+    asset_id BLOB CHECK(length(asset_id) = 32),
+
+    -- We use the 32 byte schnorr key here as this is what's used to derive the
+    -- top-level Taproot Asset commitment key.
+    group_key BLOB CHECK(LENGTH(group_key) = 32),
+    
+    leaf_node_key BLOB NOT NULL UNIQUE,
+
+    leaf_node_namespace VARCHAR NOT NULL,
+
+    -- Both the asset ID and group key cannot be null at the same time.
+    CHECK (
+        (asset_id IS NOT NULL AND group_key IS NULL) OR
+        (asset_id IS NULL AND group_key IS NOT NULL)
+    )
+);
+
+-- If there already is a multiverse root entry in the mssmt_roots for the
+-- issuance or transfer multiverses, add them to the multiverse_roots table as
+-- well. Both statements are no-ops if the root doesn't exist yet.
+INSERT INTO multiverse_roots (namespace_root, proof_type)
+SELECT 'multiverse-issuance', 'issuance'
+WHERE EXISTS (
+    SELECT 1 FROM mssmt_roots WHERE namespace = 'multiverse-issuance'
+);
+
+INSERT INTO multiverse_roots (namespace_root, proof_type)
+SELECT 'multiverse-transfer', 'transfer'
+WHERE EXISTS (
+    SELECT 1 FROM mssmt_roots WHERE namespace = 'multiverse-transfer'
+);
+
+-- And now we create the multiverse_leaves entries for the multiverse roots.
+-- This is a no-op if the multiverse root doesn't exist yet.
+INSERT INTO multiverse_leaves (
```

Comment:
This works on the overlay level, but then the underlying SMT tree wouldn't have the proper leaves/branches inserted. 

I think in addition to this, we also need a runtime/startup check where we insert the items into the tree if no actual leaves exist. We can check this with a query that properly joins against `mssmt_nodes`

Reply:
Actually this is ok. The leaves already exist in the multiverse tree, so the new join allows us to obtain that value and optimize some of the added logic (reduce number of DB transactions).